src/librustc_mir/transform/generator.rs

```diff
@@ -1015,8 +1098,20 @@ fn create_generator_resume_function<'tcx>(
 
     // Panic when resumed on the returned or poisoned state
     let generator_kind = body.generator_kind.unwrap();
-    cases.insert(1, (RETURNED, insert_panic_block(tcx, body, ResumedAfterReturn(generator_kind))));
-    cases.insert(2, (POISONED, insert_panic_block(tcx, body, ResumedAfterPanic(generator_kind))));
+
+    if can_unwind {
+        cases.insert(
+            1,
+            (POISONED, insert_panic_block(tcx, body, ResumedAfterPanic(generator_kind))),
+        );
+    }
+
+    if can_return {
+        cases.insert(
+            1,
```

Comment:
If `can_unwind` **and** `can_return`, this will overwrite the `POISONED` case.

Reply:
`insert` on a `Vec` can't overwrite.

Reply:
It should be moved to index 2 though, right?

Reply:
Ah, missed that reply

Reply:
I thought `cases` was an `FxHashMap`.